### ISSUE TITLE
Android- IOU-When user changes Stop address, the Request money amount is not updated

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#26946]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Launch app
2. Tap profile icon
3. Tap Workspaces
4. Tap any Workspace
5. Tap Reimbursements
6. Tap Track distance
7. Enter Rate and Unit
8. Tap Save
9. Navigate back to LHN
10. Tap Fab
11. Tap request money
12. Tap distance
13. Tap Start and select an address
14. Tap Finish and select an address
15. Tap Add stop
16. Tap Finish and select an address
17. Tap next
18. Select a workspace
19. Note the distance details and request money amount is correct (15.39*2=₹30.78) i. e shown as ₹30.77
20. Tap back to distance money request page
21. Tap Stop
22. Replace the entered address with a new one
23. Tap next
24. Select a Workspace
25. Note the distance field and request money amount (9.68*2=₹19.36)

## Expected Result:
When user changes Stop address, the Request money amount should be updated based on new distance.

## Actual Result:
When user changes Stop address, the Request money amount is not updated. The distance amount(15.39*2=₹30.78) before changing address is shown even after changing the address.

For new distance when address changed (9.68*2=₹19.36) , but still old amount ₹30.77 is displayed.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.65-0

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/464e7837-cbd6-40a0-ab51-1edbc6b623e8

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause-Internal Team

**Slack conversation:** @

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by ashimsharma10** [comment]
Have posted this before, 
https://expensify.slack.com/archives/C049HHMV9SM/p1693984012665899

**Comment by jeet-dhandha** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- Amount is not recalculated when the distance is changed.

### What is the root cause of that problem?
- The `shouldCalculateDistanceAmount` flag is not set to `true` when the distance is changed. due to the `props.iouAmount === 0` condition.

### What changes do you think we should make in order to solve the problem?
- We can simply remove the `shouldCalculateDistanceAmount` flag and replace it with `props.isDistanceRequest`.
- Keeping the `shouldCalculateDistanceAmount` flag doesn't make sense anymore since we are always calculating the amount when its a distance request.

- This will solve two problems:
    - The amount will be recalculated when the distance is changed.
    - The amount will be recalculated when the rate is changed.

### What alternative solutions did you explore? (Optional)
- N/A

**Comment by ahmedGaber93** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
distance - Amount isn't recalulated on changing waypoints
### What is the root cause of that problem?
In the code below, the `shouldCalculateDistanceAmount` is true only when `props.iouAmount === 0`
this is good when we calculate the amount for the first time in `const formattedAmount = `
But when we re-calculate the amount in the useEffect we stop re-calculate by it, because `shouldCalculateDistanceAmount` is true only on the first time.
https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/MoneyRequestConfirmationList.js#L173-L188
### What changes do you think we should make in order to solve the problem?
1. keep `props.iouAmount === 0` in `shouldCalculateDistanceAmount` condition, **I think it intended** because it is good to only calculate amount once here, then use `props.iouAmount` which calculated by the useEffect (will fix in step 2).

https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/MoneyRequestConfirmationList.js#L176-L179

2. replace the condition and deps in the useEffect to re-calculate the amount when any of its deps change, and the updated amount will pass to `formattedAmount` in `props.iouAmount`
```javascript
useEffect(() => {
    // if (!shouldCalculateDistanceAmount) {
    if (!props.isDistanceRequest) {
         return;
    }

    const amount = DistanceRequestUtils.getDistanceRequestAmount(distance, unit, rate);
    IOU.setMoneyRequestAmount(amount);
}, [props.isDistanceRequest, distance, rate, unit]);
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by cubuspl42** [comment]
@dukenv0307 @jeet-dhandha @ahmedGaber93 

Did you explore an alternative involving using `useMemo` (to calculate the amount) in tandem with `useEffect` (to call `setMoneyRequestAmount`), re-solving this problem cleanly? Is there some catch with that?

**Comment by cubuspl42** [comment]
@ahmedGaber93 

> keep `props.iouAmount === 0` in `shouldCalculateDistanceAmount` condition, **I think it intended**

This is a good point, [I asked here](https://github.com/Expensify/App/pull/25707/files#r1321289175)

**Comment by ahmedGaber93** [comment]
> If we keep props.iouAmount === 0 in shouldCalculateDistanceAmount, the formattedAmount that is used to display in the UI will not be updated when the distance is changed.

No, it will work fine after update `useEffect` to
```javascript
useEffect(() => {
    // if (!shouldCalculateDistanceAmount) {
    if (!props.isDistanceRequest) {
         return;
    }

    const amount = DistanceRequestUtils.getDistanceRequestAmount(distance, unit, rate);
    IOU.setMoneyRequestAmount(amount);
}, [props.isDistanceRequest, distance, rate, unit]);
```

The `useEffect` will call `IOU.setMoneyRequestAmount(amount)` which change the `props.iouAmount` that will update the `formattedAmount` without recalculating it because `shouldCalculateDistanceAmount` now is false and will use  `props.iouAmount` that updated by use effect

https://github.com/Expensify/App/blob/1eafc7dff1cb609e72a97d6f1d1993a2bf22aaa6/src/components/MoneyRequestConfirmationList.js#L177



> This is a good point, [I asked here](https://github.com/Expensify/App/pull/25707/files#r1321289175)

I remember I read before the amount can be editable manually by user in MoneyRequestConfirmationList.js, but we temporarily disable this feature.
So I think `shouldCalculateDistanceAmount` with `props.iouAmount === 0` maybe added for that, to prevent reset the amount by recalculating it after manually edit it and use `props.iouAmount`.

Let us wait the answer [here](https://github.com/Expensify/App/pull/25707/files#r1321289175).

**Comment by DylanDylann** [comment]
@cubuspl42 Re-post proposal from https://github.com/Expensify/App/issues/27051

## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Incorrect Amount after updating distance rate

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/MoneyRequestConfirmationList.js#L173
We only calculate the amount if the props.iouAmount === 0. So when the rate is changed, we don't re-calculate the amount
### What changes do you think we should make in order to solve the problem?
I think props.iouAmount === 0 condition is added to reduce the redundant computation.
In here
https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/MoneyRequestConfirmationList.js#L173-L188
we should remove the shouldCalculateDistanceAmount variable and to avoid redundant re-calculate we can create a new variable called amount in a useMemo and use it instead of calling getDistanceRequestAmount function

this is demo code change

```
const amount = useMemo(() => DistanceRequestUtils.getDistanceRequestAmount(distance, unit, rate), [distance, unit, rate]) 
const formattedAmount = CurrencyUtils.convertToDisplayString(
        props.isDistanceRequest ? amount : props.iouAmount,
        props.isDistanceRequest ? currency : props.iouCurrencyCode,
);

useEffect(() => {
        if (!props.isDistanceRequest) {
            return;
        }
        IOU.setMoneyRequestAmount(amount);
}, [props.isDistanceRequest, distance, rate, unit, amount]);

```

### What alternative solutions did you explore? (Optional)

**Comment by cubuspl42** [comment]
Hey everyone, [we've got a response](https://github.com/Expensify/App/pull/25707/files#r1322856509) from the original author of the PR. It seems that @ahmedGaber93 had good intuition here.

**Comment by ahmedGaber93** [comment]
@cubuspl42 what do you think of using Onyx store to save this value? are this option available?

```javascript
// IOU.js > add new function to set the value.
function setIsDistanceAmountOverridden(isDistanceAmountOverridden) {
    Onyx.merge(ONYXKEYS.IOU, {isDistanceAmountOverridden});
}

// IOU.js > edit this function and add isDistanceAmountOverridden rest value.
function resetMoneyRequestInfo(id = '') {
    const created = currentDate || moment().format('YYYY-MM-DD');
    Onyx.merge(ONYXKEYS.IOU, {
        ...
        isDistanceAmountOverridden: false,
    });
}


// in NewRequestAmountPage.js > navigateToNextPage
if(isDistanceRequestTab) {
    IOU.setIsDistanceAmountOverridden(true);
}

// in MoneyRequestConfirmationList.js > shouldCalculateDistanceAmount
const shouldCalculateDistanceAmount = props.isDistanceRequest && !props.isDistanceAmountOverridden;
```

<details>
<summary>result</summary>


https://github.com/Expensify/App/assets/41129870/ad71b235-fc85-454e-a273-12ad3daf7a41



</details>

**Comment by cubuspl42** [comment]
Everyone, the thread in the original PR is still open, but for now I'll assume that the safe option is to keep being open to a possibility of manual edition of the IOU amount by the user.

@ahmedGaber93 I actually marked your original proposal as non-matching the expectations; although you gave attention to the `=== 0` check, your code still would override the manually entered amount. Please correct me if I'm wrong.

**Comment by cubuspl42** [comment]
@ahmedGaber93 That sounds like something that would need to be done, but I'm actually [negotiating to postpone this work](https://github.com/Expensify/App/pull/25707/files#r1325549175) for a later time.

**Comment by DylanDylann** [comment]
@cubuspl42 Re-post proposal with new expectation

## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Incorrect Amount after updating distance rate

### What is the root cause of that problem?
Let's see here
https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/MoneyRequestConfirmationList.js#L181-L187
When coming MoneyRequestConfirmationList we will set iouAmount. Besides, we only calculate the amount if the props.iouAmount === 0. So we don't re-calculate the amount after iouAmount is set
### What changes do you think we should make in order to solve the problem?
We shouldn't update iouAmount when coming MoneyRequestConfirmationList, We only update iouAmount If user confirm the request or the user manually edits amount (it is temporary as [mentioned here](https://github.com/Expensify/App/pull/25707#discussion_r1325508766))

To do that, we need to move the code to update iouAmount 
https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/MoneyRequestConfirmationList.js#L181-L187
to confirm function
https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/MoneyRequestConfirmationList.js#L333

To safe, in MoneyRequestConfirmationList we can use DistanceRequestUtils.getDistanceRequestAmount(distance, unit, rate) as fallback of iouAmount

### Result

**Comment by cubuspl42** [comment]
@CortneyOfstad If you don't mind, I would suggesting rising this issue to $1k

I justify this by:
- we had to clear up the requirements
- some contributors already suggested multiple solutions
- the internal engineer [decided to extend the scope](https://github.com/Expensify/App/pull/25707/files#r1325553076) of this issue to handle re-addition of currently disabled feature
- what seemed like a strictly UI-related problem in the beginning we'll likely need a solution which adds new persistent state and, possibly, a beta

**Comment by DylanDylann** [comment]
@cubuspl42 Could you help to check my proposal [here](https://github.com/Expensify/App/issues/26946#issuecomment-1718973523)? As [this comment](https://github.com/Expensify/App/pull/25707/files#r1326287629), @tgolen is implementing another PR to allow the user to edit the amount. With my solution, we can fix the issue separately

**Comment by cubuspl42** [comment]
@DylanDylann Whatever the solution we go with, I would feel much more confident if we test this after the manual edition feature is re-enabled. I believe this issue is not so critical so it can't stand it.

@CortneyOfstad Could we hold this on https://github.com/Expensify/App/pull/26141?

**Comment by CortneyOfstad** [comment]
Sounds good @cubuspl42! Going to update the title and adjust the frequency of the issue from `daily` to `weekly` 👍

**Comment by hichamcc** [comment]
@izarutskaya I reported this bug first 
https://expensify.slack.com/archives/C049HHMV9SM/p1693858446228419

**Comment by tgolen** [comment]
I suggest re-evaluating the behavior of this with the changes of https://github.com/Expensify/App/issues/22715 which I am hoping to have merged today or tomorrow. A lot of this code was refactored in my PR.

**Comment by cubuspl42** [comment]
@tgolen You're right. We'll do that once it gets merged.

**Comment by tgolen** [comment]
It is merged now 👍

**Comment by DylanDylann** [comment]
Hi @tgolen @cubuspl42 we have a new blocker. Let's see my below video

https://github.com/Expensify/App/assets/141406735/45e6f0a5-693d-497e-9603-3d1e2311250b


It is caused by this PR: https://github.com/Expensify/App/pull/27797 when go to the confirmation page we add reportID into id of iou field. It makes the iou is re-set here

https://github.com/Expensify/App/blob/9795ba4451a15446a9f731d5baab2a54b11f5d78/src/libs/actions/IOU.js#L2274

So the transactionID is changed and the data is lost

**Comment by tgolen** [comment]
I think you can disregard that. All of the `shouldReset` logic is going to go away as I refactor this in https://github.com/Expensify/App/issues/26538

**Comment by DylanDylann** [comment]
@CortneyOfstad So we can hold this issue until https://github.com/Expensify/App/issues/26538 is done

**Comment by cubuspl42** [comment]
The functionality is changing under our feet. It might turn out that the bug goes away after the related PRs are gone... Still, let's track it here. Let's change the HOLD to https://github.com/Expensify/App/issues/26538.

**Comment by CortneyOfstad** [comment]
Not overdue — on hold

**Comment by cubuspl42** [comment]
@CortneyOfstad Actually, [the current HOLD status is outdated](https://github.com/Expensify/App/issues/26946#issuecomment-1740511540).

**Comment by CortneyOfstad** [comment]
Thanks for the heads up @cubuspl42! Changed the title 👍

**Comment by CortneyOfstad** [comment]
No update 👍

**Comment by cubuspl42** [comment]
FFR, we're still holding on https://github.com/Expensify/App/issues/26538

**Comment by CortneyOfstad** [comment]
No update — still waiting on https://github.com/Expensify/App/issues/26538

**Comment by CortneyOfstad** [comment]
Heading OoO until Oct. 25, so reassigning BZ to keep an eye on things while I am out 👍

**Comment by garrettmknight** [comment]
Still on hold

**Comment by CortneyOfstad** [comment]
I'm back @garrettmknight so I can take this back — thanks for holding down the fort while I was gone!

**Comment by CortneyOfstad** [comment]
Still on-hold — not overdue 👍

**Comment by CortneyOfstad** [comment]
Still on-hold. Going to change this to weekly

**Comment by cubuspl42** [comment]
For what it's worth, there's an active movement in the issue we're holding on

**Comment by CortneyOfstad** [comment]
Not overdue 👍

**Comment by cubuspl42** [comment]
(Holding on https://github.com/Expensify/App/issues/26538)

**Comment by CortneyOfstad** [comment]
Going to switch this to monthly as the priority on the holding issue is Low 👍

**Comment by DylanDylann** [comment]
@CortneyOfstad  The PR of https://github.com/Expensify/App/issues/26538 is merge.  Please remove the hold label

@cubuspl42
[Updated proposal](https://github.com/Expensify/App/issues/26946#issuecomment-1718973523)

**Comment by cubuspl42** [comment]
I approve [the proposal](https://github.com/Expensify/App/issues/26946#issuecomment-1718973523) by @DylanDylann 

I think we should ensure that the code is commented properly or even consider extracting `0` as a constant like `iouUserSetAmount`, to ensure that this logic is clear for people reading it in the future.

C+ reviewed 🎀 👀 🎀

**Comment by cubuspl42** [comment]
Our first pull request, https://github.com/Expensify/App/pull/33397, got reverted because of the regression.

@DylanDylann is working on a solution.

**Comment by CortneyOfstad** [comment]
Payment is not set until the 17th 👍

**Comment by CortneyOfstad** [comment]
Payments have been completed!

## Payment Summary

@cubuspl42 paid $250 as C+ (-50% due to regression) in Upwork
@DylanDylann paid $250 as Contributor (-50% due to regression) in Upwork

**Comment by CortneyOfstad** [comment]
@cubuspl42 any update on the checklist above? Thanks!

**Comment by ashimsharma10** [comment]
@CortneyOfstad Waiting for payment. 
I'm the issue reporter. 
Thanks.

**Comment by cubuspl42** [comment]
- The PR that introduced the bug has been identified. Link to the PR:
    -  https://github.com/Expensify/App/pull/25707
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
    - https://github.com/Expensify/App/pull/25707/files#r1322653311
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
    - No need for additional discussion
- Determine if we should create a regression test for this bug.
    - Up to the QA team
- If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.
    - Create new distance request
    - Fill start point and end point
    - Go to confirmation page and observe the amount field
    - Click back to the distance page and edit waypoint
    - Go to confirmation page and verify that the amount is updated

**Comment by CortneyOfstad** [comment]
Thanks @cubuspl42!

@ashimsharma10 any update on the proposal so I can get the payment processed? Thanks!

**Comment by ashimsharma10** [comment]
@CortneyOfstad I'm not able to apply. 
my upwork is this : https://www.upwork.com/freelancers/~018a92cf13e1e88eed

**Comment by ashimsharma10** [comment]
@Cornerstone  accepted, thank you! .
